### PR TITLE
[feat] Support different attention backends for prefill and decode 

### PR DESCRIPTION
**Pull request by Qiaolin-Yu**
## Motivation
Follow up on #6151

Note that speculative decoding is not yet supported in this PR.

## Profile

```
python3 -m sglang.bench_offline_throughput --model-path meta-llama/Llama-3.1-8B-Instruct --num-prompts 32 --random-input-len 3500 --random-output-len 25 --random-range-ratio 1 --trust-remote-code --disable-radix-cache --mem-fraction-static 0.9 --profile --reasoning-parser qwen3 --prefill-attention-backend fa3 --decode-attention-backend flashinfer
```
### Prefill (with fa3)

image

### Decode (with flashinfer)
image

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 1]
```diff
@@ -211,12 +211,25 @@ def __init__(self, model_runner: ModelRunner):
...
         self.max_bs = max(self.capture_bs)
         self.max_num_token = self.max_bs * self.num_tokens_per_bs
-        if global_server_args_dict["attention_backend"] == "flashmla":
+        if self.model_runner.decode_attn_backend is None:
+            self.decode_attn_backend = self.model_runner.attn_backend
+        else:
+            self.decode_attn_backend = self.model_runner.decode_attn_backend
+        self.decode_attn_backend.init_cuda_graph_state(self.max_num_token)
+
+        if (
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]: We can add a util function that checks current prefill_backend/decode_backend

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -953,22 +975,22 @@ def init_attention_backend(self):
...
-            self.attn_backend = TorchNativeAttnBackend(self)
-        elif self.server_args.attention_backend == "flashmla":
+            return TorchNativeAttnBackend(self)
+        elif backend_str == "flashmla":
             from sglang.srt.layers.attention.flashmla_backend import FlashMLABackend
 
-            self.attn_backend = FlashMLABackend(self)
-        elif self.server_args.attention_backend == "fa3":
+            return FlashMLABackend(self)
+        elif backend_str == "flash_attn_v3":
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/model_runner.py]: Why not "fa3" here?

File: python/sglang/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -922,24 +922,46 @@ def init_cublas(self):
         return c
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/model_runner.py]: We need a test for hybrid backend

**Comment by Fridge003** [comment]
I think a better way to implement this would be define a new `HybridBackend` class, and somehow makes it dynamically inherits methods from other backends. The implementation in this PR is too invasive and might cause many bugs.

How do you think? @Qiaolin-Yu  @zhyncs

File: python/sglang/srt/model_executor/model_runner.py (hunk 2 above)
Comment (Qiaolin-Yu) [reply python/sglang/srt/model_executor/model_runner.py]: Ohh, it's typo generated by cursor. Fixed.

**Comment by Qiaolin-Yu** [comment]
> I think a better way to implement this would be define a new `HybridBackend` class, and somehow makes it dynamically inherits methods from other backends. The implementation in this PR is too invasive and might cause many bugs.
> 
> How do you think? @Qiaolin-Yu @zhyncs

Sounds great! I think that will make a lot of sense.

File: python/sglang/srt/model_executor/model_runner.py (hunk 3 above)
Comment (Qiaolin-Yu) [reply python/sglang/srt/model_executor/model_runner.py]: I added the end-to-end test for the hybrid backend. However, it's very hard to verify if the test is actually using the correct backend in the future, since `model_runner` is more like a black box for the test layer. Do you think we should expose some functions in the `model_runner` just for test, or the current test is enough, or do you have a better idea to test it?

File: test/srt/test_hybrid_attn_backend.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import multiprocessing as mp
+import unittest
+
+import torch
+
+import sglang as sgl
+from sglang.srt.server_args import ServerArgs
+from sglang.test.runners import HFRunner, SRTRunner, check_close_model_outputs
+from sglang.test.test_utils import DEFAULT_SMALL_MODEL_NAME_FOR_TEST, CustomTestCase
+
```
Comment (Fridge003) [inline test/srt/test_hybrid_attn_backend.py:17]: We need end-to-end accuracy tests on different settings (mla, torch compile, speculative decoding etc.) You can imitate the tests in `test_fa3.py`.

But this test can also be kept.

File: python/sglang/srt/model_executor/model_runner.py (hunk 3 above)
Comment (Fridge003) [reply python/sglang/srt/model_executor/model_runner.py]: You can create a `MockModelRunner`. See `python/sglang/test/attention/test_flashattn_backend.py` for some reference.

File: python/sglang/srt/layers/attention/hybrid_attn_backend.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        k: torch.Tensor,
+        v: torch.Tensor,
+        layer: RadixAttention,
+        forward_batch: ForwardBatch,
+        save_kv_cache: bool = True,
+        **kwargs,
+    ):
+        return self.prefill_backend.forward_extend(
+            q, k, v, layer, forward_batch, save_kv_cache, **kwargs
+        )
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/hybrid_attn_backend.py:100]: Support of speculative decoding can be left for future PR.

**Comment by Fridge003** [comment]
Profile on DeepSeek-V3:
``` bash
python3 -m sglang.bench_one_batch --model /dev/shm/DeepSeek-V3/ --tp 8 --trust-remote-code --prefill-attention-backend fa3 --decode-attention-backend triton --batch 16 --input-len 1024 --output-len 2 --profile
```
FA3 kernel during prefill:
截屏2025-05-17 18 45 17

Triton kernel during decode:
截屏2025-05-17 18 45 58

Accuracy on DeepSeek-V3:
```bash
python3 -m sglang.launch_server --model /dev/shm/DeepSeek-V3/ --tp 8 --trust-remote-code --prefill-attention-backend fa3 --decode-attention-backend triton
```

```bash
Accuracy: 0.953
Invalid: 0.000
Latency: 45.295 s
Output throughput: 3147.735 token/s
```

**Comment by Fridge003** [comment]
Some benchmark results on DeepSeek-V3 TP8 on H200, with hybrid attention backends.

Benchmark command:
```bash
python3 -m sglang.bench_serving \
        --dataset-name random \
        --backend sglang-oai \
        --random-range-ratio 1 \
        --random-input-len 1024 \
        --random-output-len 1024 \
	--max-concurrency 16 \
        --num-prompts 48 \
        --model deepseek-ai/DeepSeek-V3-0324
```

| Prefill + Decode | Median TTFT (ms) | Median ITL (ms)  | Total token throughput (tok/s) |
| --------- |------------ | ------------ | ------------ |
| FA3 + FA3 | 751.17 | 19.97  |  1480.92  |
| FA3 + Triton | 752.48 | 22.43 |   1312.20 |
| FA3 + Flashinfer | 754.92 | 19.57  |   1486.53 |
| Triton + FA3 | 980.06 | 19.92 |  1471.49  |
| Flashinfer + FA3 | 781.62 | 19.98 |   1491.26  |
| Flashinfer + FlashMLA (page=64)| 796.52 | 20.53 |   1401.82 |

**Comment by zhyncs** [comment]
@Fridge003 @Qiaolin-Yu Let's add https://github.com/TianQiLin666666 as co author.

File: python/sglang/srt/model_executor/model_runner.py [hunk 6]
```diff
@@ -955,24 +955,77 @@ def init_cublas(self):
...
+
+            self.attn_backend = HybridAttnBackend(
+                decode_backend=self._get_attention_backend_from_str(
+                    self.decode_attention_backend_str
+                ),
+                prefill_backend=self._get_attention_backend_from_str(
+                    self.prefill_attention_backend_str
+                ),
+            )
+            if self.should_log:
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/model_executor/model_runner.py]: It seems that `ModelRunner` object has no attribute `should_log` @Fridge003

Comment (Fridge003) [reply python/sglang/srt/model_executor/model_runner.py]: `self.should_log` is deleted in this PR #6333

**Comment by zhyncs** [comment]
@Qiaolin-Yu @Fridge003 May you help fix the conflicts? Thanks!

**Comment by Qiaolin-Yu** [comment]
> @Qiaolin-Yu @Fridge003 May you help fix the conflicts? Thanks!

Sure.

**Comment by Edenzzzz** [comment]
Hi will this get merged anytime soon? Also, I tested on H100 to see that Flashinfer P+D beats FA3 in the mixed chunk case.

`python3 -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct --tp 2 --enable-mixed-chunk --attention-backend flashinfer`
`python3 -m sglang.bench_serving --backend sglang --dataset-name random --random-input 4000 --random-output 200 --request-rate 8 --num-prompt 480`
## Flashinfer
image

## FA3
Backend

**Comment by Qiaolin-Yu** [comment]
> Hi will this get merged anytime soon? Also, I tested on H100 to see that Flashinfer P+D beats FA3 in the mixed chunk case.
> 
> `python3 -m sglang.launch_server --model-path meta-llama/Meta-Llama-3-8B-Instruct --tp 2 --enable-mixed-chunk --attention-backend flashinfer` `python3 -m sglang.bench_serving --backend sglang --dataset-name random --random-input 4000 --random-output 200 --request-rate 8 --num-prompt 480`
> 
> ## Flashinfer
> image
> ## FA3
> Backend

Rebased now.

**Review by hnyls2002 — APPROVED** [review]

**Comment by Ximingwang-09** [comment]
Great Work. Is there any plan to support speculative decoding ?

**Comment by Fridge003** [comment]
> Great Work. Is there any plan to support speculative decoding ?

Not yet. Can you please raise an issue if you need this feature?

**Comment by Qiaolin-Yu** [comment]
> Great Work. Is there any plan to support speculative decoding ?

I can implement that when I have time. Do you need this?

**Comment by Ximingwang-09** [comment]
> > Great Work. Is there any plan to support speculative decoding ?
> 
> I can implement that when I have time. Do you need this?

Great, Thanks. But I have a question about this. When enable spec decoding, the target model's forward mode only contains prefill and target_verify, And both of them are extend mode, Do you think it's necessary to implement this feature for spec ？

**Comment by Fridge003** [comment]
> > > Great Work. Is there any plan to support speculative decoding ?
> > 
> > 
> > I can implement that when I have time. Do you need this?
> 
> Great, Thanks. But I have a question about this. When enable spec decoding, the target model's forward mode only contains prefill and target_verify, And both of them are extend mode, Do you think it's necessary to implement this feature for spec ？

Yes, there has been an issue: #9330

**Comment by Ximingwang-09** [comment]
> > > > Great Work. Is there any plan to support speculative decoding ?
> > > 
> > > 
> > > I can implement that when I have time. Do you need this?
> > 
> > 
> > Great, Thanks. But I have a question about this. When enable spec decoding, the target model's forward mode only contains prefill and target_verify, And both of them are extend mode, Do you think it's necessary to implement this feature for spec ？
> 
> Yes, there has been an issue: #9330

Ok, If needed ,maybe I'll try to fix it ？

**Comment by Fridge003** [comment]
> > > > > Great Work. Is there any plan to support speculative decoding ?
> > > > 
> > > > 
> > > > I can implement that when I have time. Do you need this?
> > > 
> > > 
> > > Great, Thanks. But I have a question about this. When enable spec decoding, the target model's forward mode only contains prefill and target_verify, And both of them are extend mode, Do you think it's necessary to implement this feature for spec ？
> > 
> > 
> > Yes, there has been an issue: #9330
> 
> Ok, If needed ,maybe I'll try to fix it ？

Cool! Contribution is highly welcomed~

**Comment by Qiaolin-Yu** [comment]
> > > > > Great Work. Is there any plan to support speculative decoding ?
> > > > 
> > > > 
> > > > I can implement that when I have time. Do you need this?
> > > 
> > > 
> > > Great, Thanks. But I have a question about this. When enable spec decoding, the target model's forward mode only contains prefill and target_verify, And both of them are extend mode, Do you think it's necessary to implement this feature for spec ？
> > 
> > 
> > Yes, there has been an issue: #9330
> 
> Ok, If needed ,maybe I'll try to fix it ？

Hi @Ximingwang-09 , I'm also working on this. Feel free to ping me on slack and we can collaborate.
